### ISSUE TITLE
[infrastructure_public_197]Delete a Job can't cleanup the completed pods it created

### ISSUE THREAD

**Issue by mdshuai** [kubernetes/kubernetes#16255]
Description of problem:
As the [job doc](https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/jobs.md#what-is-a-job) description "Deleting a Job will cleanup the pods it created."
Now when delete a job, the completed pod it created still exist.
```
Version-Release number of selected component (if applicable):
Client Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.2.320+0884214fe040e2", GitCommit:"0884214fe040e2824e9710b0a1154705122be413", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.2.320+0884214fe040e2", GitCommit:"0884214fe040e2824e9710b0a1154705122be413", GitTreeState:"clean"}

How reproducible:
Always

Steps to Reproduce:
1.Create a job
[root@ip-172-18-12-52 kubernetes]# cat job.yaml
apiVersion: extensions/v1beta1
kind: Job
metadata:
  name: pi
spec:
  completions: 5
  selector:
    matchLabels:
      app: pi
  template:
    metadata:
      name: pi
      labels:
        app: pi
    spec:
      containers:
      - name: pi
        image: perl
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
[root@ip-172-18-12-52 kubernetes]# kubectl create -f job.yaml 
job "pi" created

2.Wait the pod to be Completed
[root@ip-172-18-12-52 kubernetes]# kubectl get job
JOB       CONTAINER(S)   IMAGE(S)   SELECTOR      SUCCESSFUL
pi        pi             perl       app in (pi)   5
[root@ip-172-18-12-52 kubernetes]# kubectl get pod
NAME       READY     STATUS      RESTARTS   AGE
pi-4136c   0/1       Completed   0          4m
pi-9wwlv   0/1       Completed   0          4m
pi-e83ai   0/1       Completed   0          4m
pi-j2uxa   0/1       Completed   0          4m
pi-nnp03   0/1       Completed   0          4m

3.Delete the job
[root@ip-172-18-12-52 kubernetes]# kubectl delete job/pi
job "pi" deleted

4.Check the pod
[root@ip-172-18-12-52 kubernetes]# kubectl get pod
NAME       READY     STATUS      RESTARTS   AGE
pi-4136c   0/1       Completed   0          5m
pi-9wwlv   0/1       Completed   0          5m
pi-e83ai   0/1       Completed   0          5m
pi-j2uxa   0/1       Completed   0          5m
pi-nnp03   0/1       Completed   0          5m

Actual results:
4.The related pods still exist.

Expected results:
4.When delete job the pod should be cleanup.

Additional info:
```

**Comment by mdshuai** [comment]
@soltysh

**Comment by davidopp** [comment]
@erictune I think we need to either fix the behavior or the documentation for 1.1, do you agree?

**Comment by soltysh** [comment]
@davidopp @erictune I'm back with jobs (sorry for long absence). I'll start tackling all job related issues starting tomorrow morning.

**Comment by davidopp** [comment]
I don't think we'll be able to fix this in the 1.1 timeframe. Let's just fix the documentation.

ETA: 10/29

**Comment by erictune** [comment]
Not able to repro this.

**Comment by erictune** [comment]
Example file needs fix though.

**Comment by erictune** [comment]
Oh, example file is fixed at head.  Will try to repro with newer build

**Comment by erictune** [comment]
actually I will leave for @soltysh to repro.

**Comment by bgrant0607** [comment]
Deleting the job via the API is expected to orphan the pods.

Deleting the job via kubectl is expected to do cascading deletes on the pods by default.

What it looks like to me is that the parallelism is merely reduced to 0, which causes active pods to be killed, but doesn't clean up dead ones, as shown in the example:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/stop.go#L241
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/scale.go#L229
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/controller.go#L374

You'll need to do `kubectl get pods --show-all` to see them, though.

**Comment by soltysh** [comment]
On it, thanks @bgrant0607 for pointers.

**Comment by soltysh** [comment]
I was again influenced by the RC controller, which deals only with alive pods, but doesn't have to clean up the dead ones. I'm adding pods removal logic before removing a job. PR should be shortly.

**Comment by davidopp** [comment]
We shouldn't close this until we do the cherrypick. (I mean Kubernetes cherrypick, not the OpenShift cherrypicked mentioned elsewhere.)

**Comment by soltysh** [comment]
Just for the record, we're waiting for #16572 to close this.

**Comment by davidopp** [comment]
Fixed by #16611 (which replaced #16572)
